thundra/wrappers/cp_wrapper_utils.py

```diff
@@ -0,0 +1,206 @@
+import uuid
+import json
+import logging
+
+from thundra import constants, utils
+from thundra.application.application_info_provider import ApplicationInfoProvider
+from thundra.encoder import to_json
+from thundra.config import config_names
+from thundra.config.config_provider import ConfigProvider
+import thundra.wrappers.wrapper_utils as wrapper_utils
+
+logger = logging.getLogger(__name__)
+
+
+def on_finish(execution_context, request, response, span):
+    try:
+        if is_triggered_from_catchpoint(request):
+            on_catchpoint_request_finish(execution_context, request, response, span)
+    except Exception as e:
+        logger.error('ERROR: {}'.format(e))
+
+
+def is_triggered_from_catchpoint(request):
+    if request and request.headers:
+        headers = request.headers
+        user_agent = headers.get(constants.HTTPHeaders['USER_AGENT'])
+        return 'Catchpoint' in user_agent if user_agent else False
+    return True
+
+
+def on_catchpoint_request_finish(execution_context, request, response, span):
+    api_key = ConfigProvider.get(config_names.THUNDRA_APIKEY, None)
+    headers = request.headers if request and request.headers else {}
+    region_name = headers.get(constants.CatchpointHeaders.get('REGION_NAME'))
+    country_name = headers.get(constants.CatchpointHeaders.get('COUNTRY_NAME'))
+    city_name = headers.get(constants.CatchpointHeaders.get('CITY_NAME'))
+    test_id = headers.get(constants.CatchpointHeaders.get('TEST_ID'))
+    application_name = generate_catchpoint_application_name(region_name, country_name, city_name)
+    application_region = region_name if region_name else ''
+    trace_id = span.context.trace_id
+    transaction_id = utils.generate_id()
+    span_id = utils.generate_id()
+    start_timestamp = span.start_time
+    finish_timestamp = span.finish_time
+    duration = finish_timestamp - start_timestamp
+    application_info = get_catchpoint_application_info(application_name, application_region)
+
+    # create catchpoint invocation data
+    resource = get_catchpoint_request_resource(execution_context, request, span, duration)
+    invocation_data = create_catchpoint_request_invocation(execution_context, application_info, region_name,
+                                                           country_name, city_name, test_id, trace_id, transaction_id,
+                                                           span_id, start_timestamp, finish_timestamp, resource)
+    execution_context.report(create_report_data(api_key, 'Invocation', invocation_data))
+
+    # create catchpoint span data
+    span_data = create_catchpoint_request_span(application_info, span, resource, region_name, country_name, city_name, 
+                                               test_id, trace_id, transaction_id, span_id, start_timestamp, finish_timestamp)
+    execution_context.report(create_report_data(api_key, 'Span', span_data))
+
+    if response and hasattr(response, 'headers'):
```

Comment:
Even `response` has no headers, you need initialize headers with empty dictionary and inject `constants.THUNDRA_TRACE_ID_KEY` into headers

Reply:
done